scipy/special/cephes/stirling.c

```diff
@@ -0,0 +1,73 @@
+/* c implementation of Stirling numbers of the second kind */
+
+/*
+ *
+ *     Stirling numbers of the second kind
+ *
+ *
+ *
+ * SYNOPSIS: Stirling numbers of the second kind count the
+ *  number of ways to make a partition of n distinct elements
+ *  into k non-empty subsets.
+ *
+ * DESCRIPTION: n is the number of distinct elements of the set
+ *  to be partitioned and k is the number of non-empty subsets.
+ *  The values for n < 0 or k < 0 are interpreted as 0. If you
+ *
+ * ACCURACY: The method returns an unsigned integer type and
+ *  the size of the type 32 bits.
+ *
+ *
+ * NOTE: this file is *NOT* part of the ceph distribution and was
+ *  added by Lucas Roberts in 2023 after opening this issue
+ *  https://github.com/scipy/scipy/issues/17890
+ */
+#include "mconf.h"
+#include <stdlib.h>
+
+long stirling2(int n, int k){
+    if (n == 0 && k == 0){
+        return 1L;
+    }
+    if (k <= 0 || k > n || n <= 0){
+        return 0L;
+    }
+    int arraySize = k <= n - k + 1 ? k : n - k + 1;
+    long *curr = malloc(arraySize * sizeof(long));
+    if (!curr) {
+        sf_error("stirling2", SF_ERROR_NO_RESULT, "failed to allocate memory");
+        return -1L;
```

Comment:
```suggestion
        sf_error("stirling2", SF_ERROR_NO_RESULT, "failed to allocate memory");
        return -2L;
```

Swap to overflow -> -1, out of memory -> -2